draft-ppm-protocol.md

```diff
@@ -934,6 +961,53 @@ https://github.com/abetterinternet/prio-documents/issues/112]]
 issue#57). For example, how does a leader respond to a collect request if the
 helper drops out?]
 
+### Validating Batch Parameters {#batch-parameter-validation}
+
+Before an aggregator responds to a collect request or output share request, it
+must first check that the request does not violate the parameters associated
+with the PPM task. It does so as described here.
+
+First the aggregator checks that the request's batch interval respects the
+boundaries defined by the PPM parameters `Param`. Namely, it checks that both
+`batch_start` and `batch_end` are divisible by `Param.min_batch_duration` and
+that `batch_end - batch_window >= min_batch_duration`. Unless both these
+conditions are true, it aborts and alerts the peer with "invalid batch
+interval".
+
+Next, the aggregator checks that the request respects the generic privacy
+parameters of the PPM task. Let `X` denote the set of input shares the
+aggregator has validated and which fall in the batch interval of the request.
+
+* If `len(X) < Param.min_batch_size`, then the aggregator aborts and alerts the
+  peer with "insufficient batch size".
+* The aggregator keeps track of the number of times each input share was added
+  to the batch of an output share request. If any input share in `X` was added
```

Comment:
Interesting question. To put it another way: Is the max report lifetime too coarse-grained?

The interesting thing about Hits is that, without DP, _some amount of leakage_ is inevitable, but as you point out, how much is leaked depends on the distribution of inputs. (As an extreme example, suppose a large number of non-heavy-hitters share a common prefix.)

Vanilla Hits doesn't try to mitigate this at all, so we would set the max report lifetime to be the length of the input. (This is the maximum number of collect requests that would be needed to compute the heavy hitters.) The current proposal for mitigating this leakage is via DP. Interestingly, if the aggregators are adding the noise, then the distribution of the noise -- and therefore the privacy budget -- can only depend on the candidate prefixes of the previous rounds. I don't imagine it would, it seems reasonable to just pick a max report lifetime. But we don't want to preclude the possibility of someone doing something fancier.

Of course, nothing about the protocol precludes the aggregators from bailing out _before_ the max report lifetime is reached... what's important is that they agree on when to bail out.

Which leads me to this conclusion: the max report lifetime parameter seems to wear to hats:
1. It provides a coarse-grained way of enforcing a privacy budget. In particular, it should suffice for "simple" schemes in which the privacy budget is independent of the input distribution.
2. It provides a way for aggregators to know how long they have to keep around reports (or the output shares derived from them). Namely, if a report has reached its max lifetime, it's OK to throw it away.

This PR talks about (1.), but perhaps it also makes sense to talk about (2.).

Reply:
>Here I disagree. This operational parameter is a kind of contract between collectors and aggregators that stipulates how long data has to be available. Without it, aggregators would have to keep state indefinitely. I suppose we could do this in a different way, but we'll need to have something like this in the spec.

I'm not arguing for getting rid of this parameter, just for changing its name to better reflect what its units are (a number of queries rather than a duration). I would phrase it as "once a batch's `privacy_budget` is exhausted, aggregators MUST refuse further queries on the batch and MAY delete the inputs in the batch" (note that the MUST part is something observable by other parties whereas it's not meaningfully possible for an aggregator to prove to other participants that it has deleted some input).

But as I said, we should move forward with `max_batch_lifetime`.